rust/connlib/tunnel/src/device_channel/tun_linux.rs

```diff
@@ -0,0 +1,267 @@
+use crate::device_channel::ioctl;
+use crate::DnsFallbackStrategy;
+use connlib_shared::{messages::Interface as InterfaceConfig, Callbacks, Error, Result};
+use futures::TryStreamExt;
+use futures_util::future::BoxFuture;
+use futures_util::FutureExt;
+use ip_network::IpNetwork;
+use libc::{
+    close, fcntl, open, F_GETFL, F_SETFL, IFF_MULTI_QUEUE, IFF_NO_PI, IFF_TUN, O_NONBLOCK, O_RDWR,
+};
+use netlink_packet_route::RT_SCOPE_UNIVERSE;
+use parking_lot::Mutex;
+use rtnetlink::{new_connection, Error::NetlinkError, Handle};
+use std::task::{Context, Poll};
+use std::{
+    fmt, io,
+    os::fd::{AsRawFd, RawFd},
+};
+use tokio::io::unix::AsyncFd;
+
+mod utils;
+
+pub(crate) const SIOCGIFMTU: libc::c_ulong = libc::SIOCGIFMTU;
+
+const IFACE_NAME: &str = "tun-firezone";
+const TUNSETIFF: u64 = 0x4004_54ca;
+const TUN_FILE: &[u8] = b"/dev/net/tun\0";
+const RT_PROT_STATIC: u8 = 4;
+const DEFAULT_MTU: u32 = 1280;
+const FILE_ALREADY_EXISTS: i32 = -17;
+
+pub struct Tun {
+    handle: Handle,
+    connection: tokio::task::JoinHandle<()>,
+    fd: AsyncFd<RawFd>,
+
+    worker: Mutex<Option<BoxFuture<'static, Result<()>>>>,
```

Comment:
This `Mutex` will go away with https://github.com/firezone/firezone/pull/2659 because we won't be sharing the `Device` in an `Arc` anymore and thus can just access this `Future` with an `&mut`.